src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -141,13 +142,22 @@ data EvalRef
   = EvalExtern (Qualified Ident)
   | EvalLocal (Maybe Ident) Level
 
+data ExternQuoteInstruction
+  = QuoteAsVariable
+  | QuoteAsStop
+
+data QuoteInstruction
+  = QuoteAsExtern (Qualified Ident) ExternQuoteInstruction
+  | QuoteAsLocal (Maybe Ident) Level
+
 derive instance Eq EvalRef
 derive instance Ord EvalRef
 
 data InlineAccessor
   = InlineProp String
   | InlineSpineProp String
   | InlineRef
+  | InlineWildcard
```

Comment:
Can you clarify why this additional constructor is necessary?

Reply:
TBH I'm not sure if it's necessary, as I'm still struggling a bit to understand this ADT. There may be better ways to go about it. But my logic is as follows:

- When we store a `BackendSemantics` for dereffing, we `defer` it so that we're not doing a bunch of extra work.
- In the `eval`-ed `BackendExpr` that turns into the `BackendSemantics` over which we `deref`, we want to avoid the stack exploding in case of recursion (as we do for any `eval` on an extern), so we call `envForGroup` when we're creating this deferred term:
```purescript
giveUp = ExternPunted $ defer \_ -> evalSpine env (semanticsFromImpl (envForGroup env ref InlineWildcard (groupFromImpl impl)) qual impl) spine
```
- In the cases where we successfully case against a spine, we know what `InlineAccessor` to use in `envForGroup`. However, when the casing doesn't work, we still need an `InlineAccessor` to give to `envForGroup` to prevent the stack exploding due to a recursive call. The `InlineWildcard` constructor is intended to be this conservative backstop, as it will add an `InlineNever` to the directives that will always be picked up via `lookupOrWildcard`.

Reply:
> When we store a BackendSemantics for dereffing, we defer it so that we're not doing a bunch of extra work.

Ideally, we would only be doing conceptually constant time operations. Right now this does full evaluation on the spine (including applications), which I'm not sure we want to do. Is that relevant?

> In the eval-ed BackendExpr that turns into the BackendSemantics over which we deref, we want to avoid the stack exploding in case of recursion (as we do for any eval on an extern), so we call envForGroup when we're creating this deferred term:

I may need some examples if you have them, because I don't quite follow. To me, it seems like this work could potentially be done in `evalExtern`, with `evalExternFromImpl` remaining as-is. That is, `evalExternFromImpl` only cares about evaluating to a potentially new BackendExpr based on a spine with no other concerns. Then in `evalExpr`, it can do derefing on the input only in the case of the `Nothing` branch. The `Just` branch always yields the expression returned from `evalExternFromImpl`. Do you think there is an issue with that?

Reply:
> Ideally, we would only be doing conceptually constant time operations. Right now this does full evaluation on the spine (including applications), which I'm not sure we want to do. Is that relevant?

Full evaluation on the spine seems to be necessary in the `giveUp` function because it's not possible to incrementally evaluate the spine as we do not have context. This is in contrast to the `puntExtern` function, which uses an evaluated expression and thus doesn't need to walk the spine.

As an example, consider the following two cases:

- [Here](https://github.com/mikesol/purescript-backend-optimizer/blob/bb681c3508097a029a3e6682a2297b804f3f2ed3/src/PureScript/Backend/Optimizer/Semantics.purs#L1043), we call give up because the spine is unwalkable by `evalExternFromImpl`. If we didn't walk the spine in `giveUp`, the thing we were dereffing on would be incorrect because it wouldn't take into account the spine.
- [Here](https://github.com/mikesol/purescript-backend-optimizer/blob/bb681c3508097a029a3e6682a2297b804f3f2ed3/src/PureScript/Backend/Optimizer/Semantics.purs#L950), on the other hand, we don't need to walk the spine because it's been walked for us: the property is taken from the accessor and applied to the dictionary to get the body. In that case, we can simply give the body to `deref` and we know that we're being faithful to the spine.

In both cases, though, `puntExtern` and `giveUp` need to be lazy. This is because we don't know if we need them, and evaluation always incurs a cost. `giveUp` will almost always be more expensive because we walk the spine, but theoretically, `puntExtern` could also have a cost, for example if it winds up applying an abstraction, like it would [here](https://github.com/mikesol/purescript-backend-optimizer/blob/bb681c3508097a029a3e6682a2297b804f3f2ed3/src/PureScript/Backend/Optimizer/Semantics.purs#L920). Ofc, we want to pay that cost when we're inlining, but when we're not (which is the only time `puntExtern` is called), we want to be lazy about it.

Reply:
> I may need some examples if you have them, because I don't quite follow. To me, it seems like this work could potentially be done in evalExtern, with evalExternFromImpl remaining as-is.

From a design perspective, it very well could be the case that we'd want to do this in `evalExtern`. I don't know the architecture well enough to know what the best place is. However, I made this choice because `evalExternFromImpl` has access to an `ExternImpl` that we need to construct the `deref`fable term, whereas `evalExtern` does not have this information, meaning there's no `ExternImpl` anywhere in the signature, nor is an `ExternImpl` accessible from any of the arguments to the function.

Reply:
This makes sense, thanks, and may be a source of some of the complexity (having to do it all in `evalExternImpl`). I think it's a problem that we are not iteratively evaling the spine. Currently we have:
```purescript
evalExtern :: Env -> Qualified Ident -> Array ExternSpine -> BackendSemantics
```
That is, when we call evalExtern, there's no mention of the current ref value. But if we had:
```purescript
newtype Env = Env
  { ...
  , evalExtern :: Env -> Qualified Ident -> Array ExternSpine -> Either (Lazy BackendSemantics) BackendSemantics
  , ...
  }
evalExtern :: Env -> Qualified Ident -> Array ExternSpine -> Lazy BackendSemantics -> BackendSemantics

```
Key being that `evalExtern` takes the current ref value, and the `evalExtern` in the Env returns the `semanticsFromImpl` result in the missing case rather than "Nothing". We could then iteratively evaluate the spine in evalExtern, and `evalExternFromImpl` could stay more focused. What do you think?

Reply:
Interesting idea!

A few questions:

1. What initial value is passed as the `Lazy BackendSemantics` to `Semantics.evalExtern`?
2. Would we ignore the `Lazy BackendSemantics` produced by `env.evalExtern` in favor of the one passed into `Semantics.evalExtern`?
3. How do we track the delta spine? That is, if we call `evalExtern`, add some spine and call it again, we'd need to walk the delta spine on the current ref instead of the full spine.

Reply:
> What initial value is passed as the Lazy BackendSemantics to Semantics.evalExtern?

It would be the NeutVar I think. The initial value has no spine, so it's just the neutral reference.

> Would we ignore the Lazy BackendSemantics produced by env.evalExtern in favor of the one passed into Semantics.evalExtern

I suppose The `Lazy BackendSemantics` only applies for the initial reference (when the spine is `[]`).

> How do we track the delta spine? That is, if we call evalExtern, add some spine and call it again, we'd need to walk the delta spine on the current ref instead of the full spine.

This would be assuming that the spine is added one at a time, so it's always `Array.last` or `Array.unsnoc`. One way around this would be to have two signatures:

```purescript
-- Initial call to evaluate the reference.
evalExtern :: Env -> Qualified Ident  -> BackendSemantics

-- Subsequent calls, baking in the "unsnoc", or rather we
-- don't `snoc` before calling evalExternWithSpine.
-- Derefing is assumed to be iterative.
evalExternWithSpine :: Env -> Qualified Ident -> Array ExternSpine -> ExernSpine -> Lazy BackendSemantics -> BackendSemantics

Reply:
If I understand correctly, there are two goals:

> This makes sense, thanks, and may be a source of some of the complexity (having to do it all in evalExternImpl),

**Goal 1**: reduce the complexity of `evalExternImpl`.

> I think it's a problem that we are not iteratively evaling the spine.

**Goal 2**: iteratively eval the spine.

Goal 1 is tricky because the complexity is baked into `evalExternImpl`. It pulls the triple duty of:

1. [Getting a usable `NeutralExpr`](https://github.com/aristanetworks/purescript-backend-optimizer/blob/0927ccdcdd3e1a96c922608dea6053567b45c2dc/src/PureScript/Backend/Optimizer/Semantics.purs#L863) that we can substitute in for the variable;
2. [Walking the spine](https://github.com/aristanetworks/purescript-backend-optimizer/blob/0927ccdcdd3e1a96c922608dea6053567b45c2dc/src/PureScript/Backend/Optimizer/Semantics.purs#L895), like when we burrow into the prop of a dict; _and_
3. Deciding if we should inline it, which is expressed by the final return value: `Just x` for yea and `Nothing` for nay.

While this seems like a lot, I think it's a good thing. It follows the [Parse, Don't Validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/). As we (1) lookup the extern; (2) drill down into it using the spine; and (b) make a call if we are inlining or not, we're not just returning `true` or `false` (that'd be just "validate") but rather returning a maybe with a usable expression (so "parse").

In my PR, I leverage this by using the work done by this process in order to create the ref. That's why `puntExtern` uses the `evaled` result, whereas `giveUp` does not (because there is no result & it will return nothing), which requires us to walk the spine via `evalSpine`.

The only way I can imagine reducing the complexity of `evalExternImpl` is returning a `Tuple (Lazy BackendSemantics) (Tuple Env Boolean)` with the root expr, the env with the recursive bind group added, and `true/false` if we are inlining or no. Then we'd do the spine walk outside of the function. But I don't believe that'd simplify stuff much.

So my initial read is that the complexity of `evalExternImpl` is unavoidable given the current design. While it could be simplified by breaking it into smaller parts, in terms of what it does, I think that's sound.

***

If I understand correctly, Goal 2 is to avoid walking the spine over and over again, instead working with the diffs. But as we are only ever using the lazy value, the performance hit is constructing the closure over and over rather than walking the spine over and over. So I'm not sure if we save much time doing that.

***

So, in sum, while there are benefits to reworking these functions, I don't see clear enough gains to warrant doing it in this PR. If the PR felt sluggish I'd feel differently, but it's comparable to current `main`.

Reply:
> If I understand correctly, Goal 2 is to avoid walking the spine over and over again, instead working with the diffs. But as we are only ever using the lazy value, the performance hit is constructing the closure over and over rather than walking the spine over and over. Furthermore, baking unsnoc into the signature requires folks to avoid spine snocking before calling evalExtern, which seems fragile. So I'm not sure if we save much time doing that, and it still leaves the door open to misguided acts of snockery.

To restate, there is no advantage to iteratively evaling the spine. At no point do we observe or force an intermediate state where we might need to share previous work. We just don't know how long the spine will actually be here, and so just need to defer evaluation since it may be discarded. That makes sense.

I feel like I still need a little time to go over this because I don't understand `InlineWildcard`, and need to step through it a bit more.

Reply:
Regarding this:

> Full evaluation on the spine seems to be necessary in the giveUp function because it's not possible to incrementally evaluate the spine as we do not have context.

My point is that I don't think we should expect "deref" to perform arbitrary function evaluation right now. I know that the spine contains applications, but for the purposes of this feature, the goal is to index into data for known properties. That is, if our extern is a function, and the spine is an application, we don't want `deref` to evaluate the function because that's effectively arbitrary inlining without any guards or heuristics. This isn't necessarily a bad thing if it actually reduces to something useful, but I have no ability to reason about the effects of this in the large.

Reply:
> I feel like I still need a little time to go over this because I don't understand InlineWildcard, and need to step through it a bit more.

Absolutely, take your time, there are a lot of moving parts & I definitely don't have the full picture yet, so it's good to let things sink in.

Re `InlineWildcard`, currently, all of the inline directives correspond to something concrete, so it's an outlier. Walking through the three existing ones:

- `-- @inline foo never` is `InlineRef`
- `-- @inline foo.bar never` is `InlineProp "bar"`
- `-- @inline myInstance(..).hello` is `InlineSpineProp "hello"`

All of these are things the user can specify, but they're _also_ used internally as a backstop to recursion for the bind group. Meaning that, for example, if `foo.bar` is recursive, we insert an `InlineNever` for `InlineProp "bar"` in the directives via `envForGroup`.

In the case of `InlineWildcard`, there is no directive one can write that will produce this. We could, for example, create one. `-- @inline foo* never` would match against `foo`, `foo.bar`, `foo(..).baz`, `foo.etc.etc.etc` etc. But I'm not sure how useful that'd be. The important bit is that, as a backstop for a recursive bind group, it does this work: whenever we encounter `foo` again with any spine, it applies the directive of the wildcard. In the case of `giveUp`, this directive is `InlineNever`. Another way to think of it is: in situations where the spine is exotic, be as conservative as possible with the recursive bind groups and stop immediately (aka `Map.insert InlineWildcard InlineNever`) when an extern is hit recursively, regardless of what its spine is.

> This isn't necessarily a bad thing if it actually reduces to something useful, but I have no ability to reason about the effects of this in the large, and I don't really want to sneak that in without a clear intention of it's effects.

Agreed, I think it needs more real-world testing to understand what the effects would be. What I'm confident about is the following. Either:
- The `Lazy BackendSemantics` will reduce to something useful like an integer for `+` or an array for `Array.length`, in which case it will be used by `evalPrimOp`; or
- It will reduce to something useless like a partially applied function, in which case it will be ignored.

In terms of how that will effect compilation in the wild, I don't have any good intuition but I can offer some empirical insight. I just ran the optimizer & then this branch 5 times on the deku documentation, which is medium-sized as projects go. Here are the results:

```bash
*** current main
14.71s user 1.23s system 146% cpu 10.876 total
14.73s user 1.29s system 145% cpu 10.982 total
14.68s user 1.14s system 143% cpu 11.053 total
15.42s user 1.29s system 150% cpu 11.137 total
14.33s user 1.18s system 145% cpu 10.692 total
*** deref branch
16.16s user 1.38s system 145% cpu 12.099 total
15.15s user 1.25s system 145% cpu 11.307 total
15.92s user 1.19s system 151% cpu 11.300 total
15.76s user 1.31s system 144% cpu 11.800 total
16.30s user 1.29s system 153% cpu 11.446 total
```

So there's a performance hit of ~8%.

Reply:
> In the case of InlineWildcard, there is no directive one can write that will produce this. We could, for example, create one. -- @inline foo* never would match against foo, foo.bar, foo(..).baz, foo.etc.etc.etc etc. But I'm not sure how useful that'd be. The important bit is that, as a backstop for a recursive bind group, it does this work: whenever we encounter foo again with any spine, it applies the directive of the wildcard. In the case of giveUp, this directive is InlineNever. Another way to think of it is: in situations where the spine is exotic, be as conservative as possible with the recursive bind groups and stop immediately (aka Map.insert InlineWildcard InlineNever) when an extern is hit recursively, regardless of what its spine is.

I think that you are assuming `@inline foo never` means "when looking at an empty spine, never inline foo", which is true, but the current implementation is assuming a subsumption relationship. That is, it's assuming that a spine of n elements will always be evaluated before a spine of n+1 elements. In that case, `@inline foo never` subsumes for example `@inline foo arity=1`. `foo` with an empty spine will always be evaluated first, yielding `never`, which inserts a `NeutStop` preventing any future inlining. In that case, I think that `InlineRef` is semantically the same as `InlineWildcard`.

Reply:
Got it, we can eliminate `InlineWildcard` then 👍 Just updated the PR.